storage/src/main/java/tech/pegasys/teku/storage/store/Store.java

```diff
@@ -609,14 +640,43 @@ private SafeFuture<Optional<StateAndBlockSummary>> getOrRegenerateBlockAndState(
     if (cachedResult.isPresent()) {
       return SafeFuture.completedFuture(cachedResult);
     }
+
+    // is it an epoch boundary?
+    if (maybeEpochStates.isPresent()) {
+      final Optional<StateAndBlockSummary> maybeEpochState =
+          maybeEpochStates.get().stream()
+              .filter(
+                  stateAndBlockSummary ->
+                      stateAndBlockSummary.getBlockSummary().getRoot().equals(blockRoot))
+              .findFirst();
+      if (maybeEpochState.isPresent()) {
+        return SafeFuture.completedFuture(maybeEpochState);
+      }
+    }
     return createStateGenerationTask(blockRoot)
         .thenCompose(
             maybeTask ->
                 maybeTask.isPresent()
-                    ? states.perform(maybeTask.get())
+                    ? states.perform(maybeTask.get()).thenPeek(this::cacheIfEpochState)
                     : EmptyStoreResults.EMPTY_STATE_AND_BLOCK_SUMMARY_FUTURE);
   }
 
+  private void cacheIfEpochState(Optional<StateAndBlockSummary> maybeStateAndBlockSummary) {
+    if (maybeStateAndBlockSummary.isPresent() && maybeEpochStates.isPresent()) {
+      final UInt64 slot = maybeStateAndBlockSummary.get().getSlot();
+      final int slotsPerEpoch =
+          spec.atSlot(maybeStateAndBlockSummary.get().getSlot()).getConfig().getSlotsPerEpoch();
+      if (slot.mod(slotsPerEpoch).isZero()) {
```

Comment:
This is going to miss epochs where the first slot is empty. I suspect the key question is whether we'd want to cache the first block from the epoch or if it's better to make this a cache of checkpoint states where the state is from the last block of the previous epoch with the state transition applied. The nice thing about using that checkpoint state approach is that happens to be what we pre-cache prior to the epoch transition so we'd always generate that state and then it can be used to apply the first block of the epoch without redoing the transition regardless of how many empty slots are present at the start of the epoch. It can also be used as-is for any duty calculations that needs a state from that epoch.

But it probably depends which lookups are trying to hit this cache - it wouldn't really work for getting a state by block root in that case which I think is more what you're targeting for getting the justified/finalized states.

Reply:
given how the 'hot state persistence' is done, i might be able to more closely integrate with that - will take another pass at it...